pkg/operator/boundsatokensignercontroller/controller.go

```diff
@@ -0,0 +1,296 @@
+	signingSecret, err := c.secretClient.Secrets(operatorNamespace).Get(NextSigningKeySecretName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	needKeypair := errors.IsNotFound(err) || len(signingSecret.Data[PrivateKeyKey]) == 0 || len(signingSecret.Data[PublicKeyKey]) == 0
+	if needKeypair {
+		newSecret, err := newSigningSecret()
+		if err != nil {
+			return err
+		}
+		signingSecret, _, err = resourceapply.ApplySecret(c.secretClient, c.eventRecorder, newSecret)
+		if err != nil {
+			return err
+		}
+		// requeue for after we should have recovered
+		c.queue.AddAfter(workQueueKey, readyInterval+10*time.Second)
+	}
+
+	certConfigMap, err := c.configMapClient.ConfigMaps(targetNamespace).Get(CertConfigMapName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if errors.IsNotFound(err) {
+		certConfigMap = &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: targetNamespace,
+				Name:      CertConfigMapName,
+			},
+			Data: map[string]string{},
+		}
+	}
+	currPublicKey := string(signingSecret.Data[PublicKeyKey])
+	hasThisKey := false
+	for _, value := range certConfigMap.Data {
+		if value == currPublicKey {
+			hasThisKey = true
+			break
+		}
+	}
+	if !hasThisKey {
+		// Increment until a unique name is found
+		nextKeyIndex := len(certConfigMap.Data) + 1
+		nextKeyKey := ""
+		for len(nextKeyKey) == 0 {
+			possibleKey := fmt.Sprintf("service-account-%03d.pub", nextKeyIndex)
+			_, ok := certConfigMap.Data[possibleKey]
+			if !ok {
+				nextKeyKey = possibleKey
+				break
+			}
+			nextKeyIndex += 1
+		}
+
+		certConfigMap.Data[nextKeyKey] = currPublicKey
+		certConfigMap, _, err = resourceapply.ApplyConfigMap(c.configMapClient, c.eventRecorder, certConfigMap)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Check if next-sa-signing-key has been around long enough to be promoted.
+	// Giving time for apiserver instances to pick up the change in public keys before
+	// changing the private key minimizes the potential for one or more apiservers to
+	// issue tokens signed by the new private key that apiservers without the
+	// corresponding public key are unable to validate.
+	//
+	// TODO(marun) Find a more accurate indication that all apiservers are capable of
+	// validating tokens signed by the new private key.
+	readyToPromote := false
+	readyTime := signingSecret.Annotations[TokenReadyAnnotation]
+	if len(readyTime) == 0 {
+		readyToPromote = true
+	} else {
+		promotionTime, err := time.Parse(time.RFC3339, readyTime)
+		if err != nil || time.Now().After(promotionTime) {
+			readyToPromote = true
+		}
+	}
+
+	// If past promotion time, make the new private key available to the operand.
+	if readyToPromote {
+		_, _, err := resourceapply.SyncSecret(c.secretClient, c.eventRecorder,
+			operatorNamespace, NextSigningKeySecretName,
+			targetNamespace, SigningKeySecretName, []metav1.OwnerReference{})
+		return err
+	}
```

Comment:
should we requeue the work queue key here? Or is line 110 exactly there for that reason? (some comment here would help).

Reply:
Requeue is automatic on error. (see `processNextWorkItem`)

Reply:
No longer relevant after the refactor into separate methods.

Reply:
No longer relevent